File: backend/backing.py
"""Backing tracks -- a shelf of YouTube links you play along to.

The honest framing: this is a **bookmark list with a player attached**, and that is
deliberate. YouTube's terms allow embedding their player and driving it through the
IFrame API; they do not allow separating the audio, overlaying the video, or caching
it locally. So the two things a musician actually wants from a backing track -- pitch
shift and tempo change -- are off the table except for the one YouTube itself
provides, `setPlaybackRate`, which is a documented player control and changes pitch
with speed like a tape machine.

What is left is still worth having: a named shelf, per-track loop points so you can
grind eight bars of a solo without hunting the scrubber, a speed you set once and it
remembers, and the key and tempo you worked out written down next to the link.

Everything here is URL bookkeeping. The player is entirely in the browser -- nothing
about a backing track ever touches the audio engine, which is exactly why this module
has no reference to Engine anywhere in it.

**The one thing that will confuse people, and the reason this file knows about audio
settings at all:** in WASAPI exclusive mode Keys owns the output device, so the
browser gets silence and a backing track appears broken. The UI is told, so it can
say so rather than let you conclude the feature does not work.
"""

from __future__ import annotations

import re
import uuid
from typing import Any
from urllib.parse import parse_qs, urlparse

from . import config

MAX_TRACKS = 100

# youtu.be/<id>, /embed/<id>, /shorts/<id>, /live/<id>, /v/<id>. The watch?v= form is
# handled by the query parser instead, because a watch URL can carry the id anywhere in
# the query string.
_PATH_FORMS = re.compile(r"^/(?:embed|shorts|live|v)/([A-Za-z0-9_-]{11})")
# 11 chars of base64url is the ID format and has been since 2007.
_BARE_ID = re.compile(r"^[A-Za-z0-9_-]{11}$")


def video_id(url: str) -> str | None:
    """Pull the video id out of anything a person might paste. None if there isn't one.

    Deliberately permissive about the host and strict about the id: people paste
    music.youtube.com links, m.youtube.com links, links with a playlist and a
    timestamp and three tracking parameters, and bare ids out of a previous session.
    """
    text = (url or "").strip()
    if not text:
        return None
    if _BARE_ID.match(text):
        return text
    if "//" not in text:
        text = "https://" + text        # "youtu.be/x" pastes without a scheme

    try:
        parsed = urlparse(text)
    except ValueError:
        return None

    host = (parsed.hostname or "").lower().removeprefix("www.")
    if host not in {"youtube.com", "youtu.be", "m.youtube.com", "music.youtube.com",
                    "youtube-nocookie.com"}:
        return None

    if host == "youtu.be":
        candidate = parsed.path.lstrip("/").split("/")[0]
        return candidate if _BARE_ID.match(candidate) else None

    found = _PATH_FORMS.match(parsed.path)
    if found:
        return found.group(1)

    for value in parse_qs(parsed.query).get("v", []):
        if _BARE_ID.match(value):
            return value
    return None


def start_seconds(url: str) -> float:
    """The ?t= / &start= timestamp, if the link carries one. 0 otherwise.

    Someone sharing "the solo starts here" pastes a link with a timestamp on it, and
    throwing that away loses the only interesting thing about that particular URL.
    """
    try:
        query = parse_qs(urlparse((url or "").strip()).query)
    except ValueError:
        return 0.0
    for key in ("t", "start"):
        for raw in query.get(key, []):
            found = re.fullmatch(r"(?:(\d+)h)?(?:(\d+)m)?(?:(\d+)s?)?", raw.strip())
            if found and any(found.groups()):
                hrs, mins, secs = (int(g or 0) for g in found.groups())
                return float(hrs * 3600 + mins * 60 + secs)
    return 0.0


class Backing:
    """The shelf. Persisted in config.local.json alongside every other preference."""

    def __init__(self, settings: config.Settings | None = None) -> None:
        self.settings = settings or config.settings

    def all(self) -> list[dict[str, Any]]:
        raw = self.settings.get("backing", "tracks", default=[]) or []
        return [t for t in raw if isinstance(t, dict) and t.get("video")]

    def _save(self, tracks: list[dict[str, Any]]) -> list[dict[str, Any]]:
        # Settings deep-merges dicts but replaces lists wholesale, which is what we
        # want: the client never sends a patch to one track, it sends the shelf.
        self.settings.update({"backing": {"tracks": tracks[:MAX_TRACKS]}})
        return tracks[:MAX_TRACKS]

    def add(self, url: str, title: str = "") -> tuple[list[dict[str, Any]], str]:
        vid = video_id(url)
        if vid is None:
            return self.all(), "that does not look like a YouTube link"
        tracks = self.all()
        if any(t["video"] == vid for t in tracks):
            return tracks, "that track is already on the shelf"
        if len(tracks) >= MAX_TRACKS:
            return tracks, f"the shelf holds {MAX_TRACKS} tracks"
        tracks.append({
            "id": uuid.uuid4().hex[:8],
            "video": vid,
            "title": (title or "").strip()[:120] or vid,
            "url": url.strip()[:400],
            "key": "",
            "bpm": 0,
            "notes": "",
            "rate": 1.0,
            # Loop points in seconds. b <= a means "no loop", which is why they start
            # equal rather than at some sentinel.
            "loop_a": start_seconds(url),
            "loop_b": start_seconds(url),
        })
        return self._save(tracks), ""

    def update(self, track_id: str, patch: dict[str, Any]) -> tuple[list[dict[str, Any]], str]:
        tracks = self.all()
        found = next((t for t in tracks if t.get("id") == track_id), None)
        if found is None:
            return tracks, "no such track"
        if "title" in patch:
            found["title"] = str(patch["title"]).strip()[:120] or found["video"]
        if "key" in patch:
            found["key"] = str(patch["key"]).strip()[:12]
        if "notes" in patch:
            found["notes"] = str(patch["notes"]).strip()[:300]
        if "bpm" in patch:
            try:
                found["bpm"] = max(0, min(400, int(float(patch["bpm"]))))
            except (TypeError, ValueError):
                found["bpm"] = 0
        if "rate" in patch:
            # YouTube's own range. Anything outside it is silently ignored by the
            # player, which would read as the control being broken.
            found["rate"] = max(0.25, min(2.0, float(patch["rate"])))
        for key in ("loop_a", "loop_b"):
            if key in patch:
                found[key] = max(0.0, float(patch[key]))
        return self._save(tracks), ""

    def remove(self, track_id: str) -> list[dict[str, Any]]:
        return self._save([t for t in self.all() if t.get("id") != track_id])

File: frontend/backing.js
/* Backing tracks -- a shelf of YouTube links, with a player attached.
 *
 * Two constraints shape everything here.
 *
 * **The audio device.** In WASAPI exclusive mode Keys owns the output and the browser
 * gets silence, so a backing track looks broken when it is merely blocked. The panel
 * checks, says so, and offers the one-click switch rather than letting you conclude
 * the feature does not work.
 *
 * **YouTube's terms.** Embedding their player and driving it through the IFrame API is
 * allowed; separating the audio, overlaying the video and caching it are not. So the
 * two things a musician wants -- pitch shift and independent tempo -- are off the
 * table, except for setPlaybackRate, which is a documented player control and moves
 * pitch with speed like a tape machine. Slow practice still works; it just transposes.
 *
 * The API script is fetched the first time you open a track and never before. Keys
 * makes no network request until you ask it to, and this is the only place it can.
 */

import { onPanic } from './app.js';
import { $, api, h, hms, mod, slider, toast } from './ui.js';

const API_SRC = 'https://www.youtube.com/iframe_api';
const RATES = [0.5, 0.75, 0.9, 1, 1.25, 1.5];

let apiReady = null;

/* One promise for the whole page. YT calls a single global when it loads, so a second
   concurrent request has to wait on the first rather than overwrite the callback. */
function loadApi() {
  if (apiReady) return apiReady;
  apiReady = new Promise((resolve, reject) => {
    if (window.YT?.Player) { resolve(window.YT); return; }
    const prev = window.onYouTubeIframeAPIReady;
    window.onYouTubeIframeAPIReady = () => { prev?.(); resolve(window.YT); };
    const tag = document.createElement('script');
    tag.src = API_SRC;
    tag.onerror = () => reject(new Error('could not reach YouTube -- are you online?'));
    document.head.append(tag);
  });
  return apiReady;
}

export function createBacking() {
  let offPanic = null;          // unsubscribes this player from PANIC
  let tracks = [];
  let exclusive = true;
  let player = null;
  let currentId = null;
  let watch = null;          // the A/B loop watcher
  let listSig = null;

  const el = mod('Backing tracks', 'play along with something',
    h('div.note', null,
      'Paste a YouTube link and it goes on the shelf. Set loop points to grind eight ',
      'bars of a solo without hunting the scrubber, slow it down to learn it, and write ',
      'the key and tempo next to it so you do not work them out twice.'),

    // Shown only if the automatic hand-off in open() failed. Normally you never see it.
    h('div.note.note--warn', { id: 'yt-excl', style: { display: 'none' } },
      h('strong', null, 'Keys still owns the speakers.'),
      ' Exclusive mode is where the 3 ms comes from, and while it holds the output ',
      'device nothing else can make a sound -- including this video. ',
      h('button.btn', { id: 'yt-share', style: { marginLeft: '8px' } }, 'Switch to shared'),
      ' Latency goes to roughly 10 ms, which is still very playable.'),

    // The state you are actually in while using this panel.
    h('div.note', { id: 'yt-shared', style: { display: 'none' } },
      h('strong', null, 'Shared output'), ' -- the video and your piano are both audible. ',
      'Latency is roughly 10 ms instead of 3. ',
      h('button.btn', { id: 'yt-restore', style: { marginLeft: '8px' } },
        'Give the device back to Keys (3 ms)')),

    h('div.btnrow', { style: { margin: '12px 0' } },
      h('input', {
        type: 'text', id: 'yt-url', placeholder: 'paste a YouTube link',
        style: { flex: '1', minWidth: '220px' },
        onkeydown: (e) => { if (e.key === 'Enter') add(); },
      }),
      h('button.btn', { onclick: add }, 'Add to shelf')),

    h('div.yt', { id: 'yt-stage', style: { display: 'none' } },
      h('div.yt__frame', null, h('div', { id: 'yt-player' })),
      h('div.yt__side', null,
        h('div.yt__title', { id: 'yt-title' }, ''),
        h('div.note', null,
          'The player has YouTube\'s own controls -- play, pause, scrub, volume, ',
          'quality, fullscreen. The buttons here are the ones it does not have.'),
        h('div.btnrow', null,
          h('button.btn', { id: 'yt-play' }, 'Play'),
          h('button.btn', { id: 'yt-a' }, 'Set A'),
          h('button.btn', { id: 'yt-b' }, 'Set B'),
          h('button.btn', { id: 'yt-clear' }, 'Clear loop'),
          h('button.btn', { id: 'yt-close' }, 'Close')),
        h('div.yt__loop', { id: 'yt-loop' }, 'no loop set'),
        h('label.field', null,
          h('span.field__label', null, h('span', null, 'Speed'),
            h('span.field__value', { id: 'yt-rate-v' }, '1x')),
          h('div.btnrow', { id: 'yt-rates' },
            RATES.map((r) => h('button.btn', {
              'data-rate': r, onclick: () => setRate(r),
            }, r === 1 ? '1x' : `${r}x`)))),
        h('div.note', null,
          'YouTube changes pitch with speed, like slowing a record. At 0.75x everything ',
          'sounds a fourth or so low -- fine for learning the shapes, not for playing ',
          'along in the original key.'))),

    h('div', { id: 'yt-list' }));

  /* ── shelf ──────────────────────────────────────────────────────────────── */
  function apply(res) {
    if (!res) return;
    if (res.tracks) tracks = res.tracks;
    if ('exclusive' in res) exclusive = res.exclusive;
    if (res.error) toast(res.error, 'bad');
    paint();
  }

  async function add() {
    const url = $('#yt-url').value.trim();
    if (!url) return;
    try {
      const res = await api.post('/api/backing', { url });
      apply(res);
      if (!res.error) { $('#yt-url').value = ''; toast('On the shelf', 'good'); }
    } catch (err) { toast(err.message, 'bad'); }
  }

  const patch = (id, body) => api.post(`/api/backing/${id}`, body)
    .then(apply).catch((err) => toast(err.message, 'bad'));

  async function remove(id) {
    if (currentId === id) closePlayer();
    try { apply(await api.del(`/api/backing/${id}`)); }
    catch (err) { toast(err.message, 'bad'); }
  }

  function paint() {
    const open_ = !!currentId;
    // Null-guarded throughout: reclaimDevice() can resolve after the view has been
    // unmounted and every one of these nodes has been thrown away.
    const excl = $('#yt-excl');
    if (!excl) return;
    // The warning is for the case the hand-off failed; the shared notice is for the
    // normal one. Neither is worth showing until there is a track open to hear.
    excl.style.display = exclusive && open_ ? '' : 'none';
    $('#yt-shared').style.display = !exclusive && open_ ? '' : 'none';

    const sig = tracks.map((t) => `${t.id}:${t.title}:${t.key}:${t.bpm}:${t.loop_a}:${t.loop_b}`).join('|');
    if (sig === listSig) return;
    listSig = sig;

    const host = $('#yt-list');
    if (!tracks.length) {
      host.replaceChildren(h('div.empty', null, 'nothing on the shelf yet'));
      return;
    }
    host.replaceChildren(...tracks.map((t) => h('div.track', {
      id: 'tr-' + t.id, class: t.id === currentId ? 'is-on' : '',
    },
      h('div.track__head', null,
        h('button.btn', { onclick: () => open(t) }, t.id === currentId ? 'Open' : 'Play'),
        h('input.track__title', {
          type: 'text', value: t.title,
          onchange: (e) => patch(t.id, { title: e.target.value }),
        }),
        looped(t) ? h('span.tag.tag--amber', null,
          `loop ${hms(t.loop_a)}-${hms(t.loop_b)}`) : null,
        h('span.list__spacer'),
        h('input.track__meta', {
          type: 'text', value: t.key, placeholder: 'key',
          onchange: (e) => patch(t.id, { key: e.target.value }),
        }),
        h('input.track__meta', {
          type: 'number', value: t.bpm || '', placeholder: 'bpm',
          onchange: (e) => patch(t.id, { bpm: e.target.value || 0 }),
        }),
        h('button.btn', { onclick: () => remove(t.id) }, 'Remove')),
      h('input.track__notes', {
        type: 'text', value: t.notes, placeholder: 'notes -- what you are working on here',
        onchange: (e) => patch(t.id, { notes: e.target.value }),
      }))));
  }

  const looped = (t) => (t.loop_b || 0) - (t.loop_a || 0) > 0.5;
  const track = (id) => tracks.find((t) => t.id === id);

  /* Hand the output device over so the video can actually be heard.
   *
   * WASAPI exclusive is first-come and total: while Keys holds the endpoint, nothing
   * else on that device makes a sound. A backing track that plays in silence is not a
   * backing track, so opening one trades 3 ms for ~10 ms automatically rather than
   * making you go and find a setting first. It says so, and hands it straight back
   * on one click or when you close the track. */
  async function yieldDevice() {
    if (!exclusive) return true;
    try {
      const res = await api.post('/api/audio', { exclusive: false });
      exclusive = !!res.engine?.exclusive;
      if (!exclusive) {
        toast('Shared output -- the video has sound now, and so does your piano', 'good', 5000);
      }
      for (const w of res.warnings || []) toast(w, 'bad', 8000);
    } catch (err) {
      toast(`Could not release the audio device: ${err.message}`, 'bad', 8000);
    }
    paint();
    return !exclusive;
  }

  async function reclaimDevice() {
    try {
      const res = await api.post('/api/audio', { exclusive: true, period_size: 144 });
      exclusive = !!res.engine?.exclusive;
      toast(exclusive ? 'Exclusive again -- 3.00 ms. Video audio is muted while it holds.'
                      : 'Could not reacquire the device', exclusive ? 'good' : 'bad', 5000);
    } catch (err) { toast(err.message, 'bad'); }
    paint();
  }

  /* ── player ─────────────────────────────────────────────────────────────── */
  async function open(t) {
    $('#yt-stage').style.display = '';
    $('#yt-title').textContent = t.title;
    currentId = t.id;
    listSig = null;
    paint();
    markRate(t.rate || 1);
    showLoop(t);

    // Before the player exists, so the first frame of audio has somewhere to go.
    // The engine reopens its stream here, which takes a moment -- doing it mid-playback
    // would cut the video off a second after it started.
    await yieldDevice();

    let YT;
    try { YT = await loadApi(); }
    catch (err) { toast(err.message, 'bad', 8000); return; }

    if (player) {
      player.loadVideoById({ videoId: t.video, startSeconds: t.loop_a || 0 });
      player.setPlaybackRate(t.rate || 1);
      return;
    }
    player = new YT.Player('yt-player', {
      videoId: t.video,
      // origin is required for the API to talk to the frame at all; 127.0.0.1 is a
      // legitimate origin as far as the player is concerned.
      playerVars: { origin: location.origin, rel: 0, modestbranding: 1,
                    start: Math.floor(t.loop_a || 0) },
      events: {
        onReady: () => { player.setPlaybackRate(t.rate || 1); startWatch(); },
        onStateChange: (e) => {
          const btn = $('#yt-play');
          if (btn) btn.textContent = e.data === 1 ? 'Pause' : 'Play';
        },
        onError: (e) => playerError(e.data, track(currentId)),
      },
    });
  }

  /* Plenty of music is embed-blocked by the rights holder, and the player's own
     "Video unavailable" card does not say which of the several reasons applies. It
     reads as the app being broken, which it is not -- the link is simply not one you
     can play here. */
  function playerError(code, t) {
    const why = {
      2: 'that video id is malformed',
      5: 'the player could not start -- try reloading the page',
      100: 'that video is private, deleted, or does not exist',
      101: 'the owner does not allow this one to be played outside YouTube',
      150: 'the owner does not allow this one to be played outside YouTube',
    }[code] || `the player returned error ${code}`;
    toast(`Cannot play it here: ${why}`, 'bad', 9000);
    const host = $('#yt-loop');
    if (host && t) {
      host.replaceChildren(
        h('span', null, why, ' '),
        h('a', { href: t.url || `https://youtu.be/${t.video}`, target: '_blank',
                 rel: 'noreferrer' }, 'open it on YouTube'));
    }
  }

  function closePlayer() {
    stopWatch();
    currentId = null;
    listSig = null;
    if (player?.stopVideo) { try { player.stopVideo(); } catch { /* already gone */ } }
    $('#yt-stage').style.display = 'none';
    paint();
    // Closing the track is the clearest possible statement that you are done needing
    // the video audible, so the low-latency path comes back without being asked.
    if (!exclusive) reclaimDevice();
  }

  /* The A/B loop. A 10 Hz poll on a video scrubber, which is nothing like musical
     timing -- the rule against setInterval is about notes, and this is a seek. */
  function startWatch() {
    stopWatch();
    watch = setInterval(() => {
      const t = track(currentId);
      if (!t || !player?.getCurrentTime) return;
      const now = player.getCurrentTime();
      const el2 = $('#yt-now');
      if (el2) el2.textContent = hms(now);
      if (looped(t) && now >= t.loop_b) player.seekTo(t.loop_a, true);
    }, 100);
  }

  function stopWatch() {
    if (watch) clearInterval(watch);
    watch = null;
  }

  function showLoop(t) {
    const host = $('#yt-loop');
    if (!host) return;
    host.replaceChildren(
      looped(t)
        ? h('span', null, 'looping ', h('strong', null, hms(t.loop_a)), ' to ',
            h('strong', null, hms(t.loop_b)), ' · at ')
        : h('span', null, 'no loop set · at '),
      h('span', { id: 'yt-now' }, '0:00'));
  }

  function setRate(r) {
    const t = track(currentId);
    if (!t) return;
    player?.setPlaybackRate?.(r);
    markRate(r);
    patch(t.id, { rate: r });
  }

  function markRate(r) {
    $('#yt-rate-v').textContent = r === 1 ? '1x' : `${r}x`;
    for (const b of document.querySelectorAll('#yt-rates .btn')) {
      b.classList.toggle('is-on', Number(b.dataset.rate) === r);
    }
  }

  function wire() {
    $('#yt-play').onclick = () => {
      if (!player) return;
      if (player.getPlayerState?.() === 1) player.pauseVideo();
      else player.playVideo();
    };
    $('#yt-a').onclick = () => mark('loop_a');
    $('#yt-b').onclick = () => mark('loop_b');
    $('#yt-clear').onclick = () => {
      const t = track(currentId);
      if (t) patch(t.id, { loop_a: 0, loop_b: 0 }).then(() => showLoop(track(currentId)));
    };
    $('#yt-share').onclick = yieldDevice;
    $('#yt-restore').onclick = reclaimDevice;
    $('#yt-close').onclick = closePlayer;
  }

  function mark(which) {
    const t = track(currentId);
    if (!t || !player?.getCurrentTime) return;
    const at = player.getCurrentTime();
    // Setting A past B is a mistake, not an instruction to play backwards.
    const body = which === 'loop_a'
      ? { loop_a: at, loop_b: Math.max(t.loop_b || 0, at) }
      : { loop_b: at, loop_a: Math.min(t.loop_a || 0, at) };
    patch(t.id, body).then(() => { listSig = null; paint(); showLoop(track(currentId)); });
  }

  return {
    el,
    async init() {
      wire();
      /* PANIC has to reach this. Everything else Keys can silence goes through the
         synth; a video does not, and "make it stop" that leaves a backing track
         playing is not what the button says. Returns whether it stopped anything so
         the toast can name it. */
      offPanic = onPanic(() => {
        if (player?.getPlayerState?.() !== 1) return false;
        player.pauseVideo();
        return true;
      });
      try { apply(await api.get('/api/backing')); }
      catch { /* the shelf is a preference list; the app works without it */ }
    },
    status(s) {
      const now = !!s.engine?.exclusive;
      if (now !== exclusive) { exclusive = now; paint(); }
    },
    destroy() {
      stopWatch();
      offPanic?.();
      offPanic = null;
      // Leaving Tools throws the iframe away with the rest of the view, so the video
      // is gone whether or not you meant it to be -- do not leave the audio device in
      // shared mode paying 7 ms for a player that no longer exists.
      const wasOpen = !!currentId;
      player = null;
      currentId = null;
      listSig = null;
      if (wasOpen && !exclusive) {
        api.post('/api/audio', { exclusive: true, period_size: 144 }).catch(() => {});
      }
    },
  };
}
